diff --git a/data/maps/earth/openstreetmap/openstreetmap.dgml b/data/maps/earth/openstreetmap/openstreetmap.dgml
index c3ebe94..2fb6d0e 100644
--- a/data/maps/earth/openstreetmap/openstreetmap.dgml
+++ b/data/maps/earth/openstreetmap/openstreetmap.dgml
@@ -17,7 +17,7 @@
       <visible>true</visible>
       <description><![CDATA[<p>A <i>global roadmap</i> created by the OpenStreetMap (OSM) project.</p><p>OSM is an open community which creates free editable maps. The OSM data was rendered using Mapnik.</p>]]></description>
       <zoom>
-        <minimum>   900  </minimum>
+        <minimum>  1400  </minimum>
         <maximum>  3500  </maximum>
         <discrete> true </discrete>
       </zoom>
@@ -31,7 +31,7 @@
         <texture name="mapnik_data" expire="604800">
           <sourcedir format="PNG"> earth/openstreetmap </sourcedir>
           <tileSize width="256" height="256"/>
-          <storageLayout levelZeroColumns="1" levelZeroRows="1" maximumTileLevel="19" mode="OpenStreetMap"/>
+          <storageLayout levelZeroColumns="1" levelZeroRows="1" maximumTileLevel="20" mode="OpenStreetMap"/>
           <projection name="Mercator"/>
           <downloadUrl protocol="https" host="a.tile.openstreetmap.org" path="/"/>
           <downloadUrl protocol="https" host="b.tile.openstreetmap.org" path="/"/>
diff --git a/src/lib/marble/CMakeLists.txt b/src/lib/marble/CMakeLists.txt
index a47b568..5c862f0 100644
--- a/src/lib/marble/CMakeLists.txt
+++ b/src/lib/marble/CMakeLists.txt
@@ -378,6 +378,7 @@ target_link_libraries(marblewidget
     LINK_PUBLIC
         Qt5::Xml
         Qt5::Gui
+	Qt5::Quick
         Qt5::Widgets
         Qt5::Network
     LINK_PRIVATE
@@ -562,8 +563,32 @@ install( FILES
     geodata/parser/GeoDocument.h
     geodata/parser/GeoDataTypes.h
     geodata/writer/GeoWriter.h
+    geodata/scene/GeoSceneIcon.h
+    geodata/scene/GeoSceneTileDataset.h
+    geodata/scene/GeoSceneVectorTileDataset.h
+    geodata/scene/GeoSceneGeodata.h
+    geodata/scene/GeoSceneGroup.h
+    geodata/scene/GeoSceneZoom.h
+    geodata/scene/GeoSceneLegend.h
+    geodata/scene/GeoSceneTextureTileDataset.h
+    geodata/scene/GeoSceneAbstractDataset.h
+    geodata/scene/GeoSceneItem.h
+    geodata/scene/GeoSceneLicense.h
+    geodata/scene/GeoSceneSection.h
+    geodata/scene/GeoSceneFilter.h
+    geodata/scene/GeoSceneHead.h
+    geodata/scene/GeoSceneVector.h
+    geodata/scene/GeoSceneSettings.h
+    geodata/scene/GeoSceneDocument.h
+    geodata/scene/GeoSceneMap.h
+    geodata/scene/GeoSceneProperty.h
+    geodata/scene/GeoSceneLayer.h
+    geodata/scene/GeoScenePalette.h
     routing/RoutingWidget.h
     routing/RoutingManager.h
+    declarative/MarbleQuickItem.h
+    declarative/Placemark.h
+    declarative/Coordinate.h
     TileCreator.h
     PluginManager.h
     PluginInterface.h
diff --git a/src/lib/marble/PlacemarkLayout.cpp b/src/lib/marble/PlacemarkLayout.cpp
index 0677983..671587d 100644
--- a/src/lib/marble/PlacemarkLayout.cpp
+++ b/src/lib/marble/PlacemarkLayout.cpp
@@ -305,6 +305,7 @@ void PlacemarkLayout::addPlacemarks( const QModelIndex& parent, int first, int l
         TileId key = TileId::fromCoordinates( coordinates, zoomLevel );
         m_placemarkCache[key].append( placemark );
     }
+    requestStyleReset();
     emit repaintNeeded();
 }
 
diff --git a/src/lib/marble/declarative/MarbleQuickItem.h b/src/lib/marble/declarative/MarbleQuickItem.h
index 151bc39..c528030 100644
--- a/src/lib/marble/declarative/MarbleQuickItem.h
+++ b/src/lib/marble/declarative/MarbleQuickItem.h
@@ -11,7 +11,6 @@
 #ifndef MARBLEQUICKITEM_H
 #define MARBLEQUICKITEM_H
 
-#include "marble_declarative_export.h"
 #include <QSharedPointer>
 #include <QQuickPaintedItem>
 #include "GeoDataAccuracy.h"
@@ -30,7 +29,7 @@ namespace Marble
     class MarbleQuickItemPrivate;
 
     //Class is still being developed
-    class MARBLE_DECLARATIVE_EXPORT MarbleQuickItem : public QQuickPaintedItem
+    class Q_DECL_EXPORT MarbleQuickItem : public QQuickPaintedItem
     {
     Q_OBJECT
 
diff --git a/src/lib/marble/geodata/scene/GeoSceneTextureTileDataset.h b/src/lib/marble/geodata/scene/GeoSceneTextureTileDataset.h
index 8bd9437..835b91e 100644
--- a/src/lib/marble/geodata/scene/GeoSceneTextureTileDataset.h
+++ b/src/lib/marble/geodata/scene/GeoSceneTextureTileDataset.h
@@ -21,7 +21,7 @@
 namespace Marble
 {
 
-class GeoSceneTextureTileDataset : public GeoSceneTileDataset
+class GEODATA_EXPORT GeoSceneTextureTileDataset : public GeoSceneTileDataset
 {
  public:
 
diff --git a/src/plugins/CMakeLists.txt b/src/plugins/CMakeLists.txt
index cb5915f..e4ad30b 100644
--- a/src/plugins/CMakeLists.txt
+++ b/src/plugins/CMakeLists.txt
@@ -3,6 +3,6 @@ if(NOT CMAKE_SYSTEM_NAME STREQUAL Android)
     add_subdirectory( templates )
 endif()
 
-add_subdirectory( render )
-add_subdirectory( positionprovider )
-add_subdirectory( runner )
+#add_subdirectory( render )
+#add_subdirectory( positionprovider )
+#add_subdirectory( runner )
